File: backend/app/schemas/run_schemas.py
from typing import Optional, Dict, Any, List
from pydantic import BaseModel
from datetime import datetime

from .workflow_schemas import WorkflowVersionResponseSchema
from ..models.run_model import RunStatus
from .task_schemas import TaskResponseSchema


class StartRunRequestSchema(BaseModel):
    initial_inputs: Optional[Dict[str, Dict[str, Any]]] = None
    parent_run_id: Optional[str] = None


class RunResponseSchema(BaseModel):
    id: str
    workflow_id: str
    workflow_version_id: int
    workflow_version: WorkflowVersionResponseSchema
    status: RunStatus
    run_type: str
    initial_inputs: Optional[Dict[str, Dict[str, Any]]]
    input_dataset_id: Optional[str]
    outputs: Optional[Dict[str, Any]]
    output_file_id: Optional[str]
    start_time: Optional[datetime]
    end_time: Optional[datetime]
    tasks: List[TaskResponseSchema]

    class Config:
        from_attributes = True


class PartialRunRequestSchema(BaseModel):
    node_id: str
    rerun_predecessors: bool = False
    initial_inputs: Optional[Dict[str, Dict[str, Any]]] = None
    partial_outputs: Optional[Dict[str, Dict[str, Any]]] = None


class BatchRunRequestSchema(BaseModel):
    dataset_id: str
    mini_batch_size: int = 10

File: frontend/src/types/api_types/runSchemas.ts
import { WorkflowVersionResponse } from './workflowSchemas';
import { TaskResponse } from './taskSchemas';

export type RunStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED' | 'CANCELLED';

export interface StartRunRequest {
  initial_inputs?: Record<string, Record<string, any>>;
  parent_run_id?: string;
}

export interface RunResponse {
  id: string;
  workflow_id: string;
  workflow_version_id: number;
  workflow_version: WorkflowVersionResponse;
  status: RunStatus;
  run_type: string;
  initial_inputs?: Record<string, Record<string, any>>;
  input_dataset_id?: string;
  outputs?: Record<string, any>;
  output_file_id?: string;
  start_time?: string;
  end_time?: string;
  tasks: TaskResponse[];
}

export interface PartialRunRequest {
  node_id: string;
  rerun_predecessors: boolean;
  initial_inputs?: Record<string, Record<string, any>>;
  partial_outputs?: Record<string, Record<string, any>>;
}

export interface BatchRunRequest {
  dataset_id: string;
  mini_batch_size: number;
}
